File: module 2/part c/Answers-BeginnerToModerate.py
#**********************
# Variables in Python
#**********************

#E1
# Declare a variable (i.e. something that stores a value) and initialize it (i.e. set something to that variable)
someVariable = 0
print(someVariable)

#E2
# Change the value of that variable (i.e. change what was initialized before), go ahead and try 
# to reset/declare, that variable a few times, check the variable type, (use type function, i.e. is 
# it int, string, etc.)
# Note: Run each variable declaration separately
someVariable = "foobar"
someVariable = "foo"
someVariable = 111

#(remember run print for each statement separately)
print(someVariable)
print(type(someVariable))

#E3
# Global vs. local variables in functions
def someFunction():
  #this variable is LOCAL only
  someOtherVariable = "def"
  print(someOtherVariable)

someFunction()
# you should get 'NameError: name 'someOtherVariable' is not defined' (variable ONLY exists in the function)
print(someVariable) 

# This is the variable we declared earlier (it's GLOBAL)
print(someVariable)


#E4 
# Pass function to a function
def someOtherFunc(arg):
    arg=arg+1
    return arg

def someFunc(someOtherFunc):
    arg = someOtherFunc(5)
    print(arg)

#E5
#Now delete all the variables we made
del someVariable

#**********************
# Conditions in Python
#**********************

#E6
# Write a conditional in Python
def compare():
  a, b = 10, 100
  
  # conditional flow uses if, elif, else (no case statements in python) 
  if(a < b):
    result= "a is less than b"
  elif (a == b):
    result= "a is same as b"
  else:
    result= "a is greater than b"
  print(result)

#**********************
# Functions in Python
#**********************

#E7
# Define a function (i.e. takes inputs, outputs, in this case a function with NO arguments,
# NO return values)
def someFunction():
  print(1+1)

#E8
# Define a function (i.e. takes inputs, outputs, in this case a function with arguments)
# NO return values)
def someFunction(a,b):
  print(a+b)

#E9
# Define a function (i.e. takes inputs, outputs, in this case a function with arguments)
# return values)
def someFunction(a,b):
  print(a+b)
  return (a+b)

#E10
### Advanced
# Define a function (i.e. takes inputs, outputs, in this case a function with arguments)
# return values, has DEFAULT values and VARIABLE number of arguments)
def someFunction(a=1,b=1,*args):
  total = 0;
  #Multiple arg's
  for i in args:
    total = a + b + i + total
  return total

someFunction(1,1,4,5,3,3)

#**********************
# Loops in Python
#**********************

#E11
# Create a while loop
def someFunction():
  x = 0
  
  # create a while loop
  while (x < 5):
     print(x)
     x = x + 1

  # create a for loop
  for x in range(5,10):
    print(x)
  
#E12
# Create a for loop
def someFunction():
  x = 0

  # create a for loop
  for x in range(5,10):
    print(x)

#E13
### Advanced
# loop over a collection, use the enumerate() function to get index, and break and continue (generally)
# don't want to 'break' out of loops

def someFunction():

  # use a for loop over a collection
  countries = ["Brazil","Turkey","Iran","Canada","Nigeria","England","Vietnam"]
  for c in countries:

  	print ("Before breaking " + c)
  	if c == "Canada": 
  		break
  	if c != "Japan": 
  		continue
  	print(c)

  for i, c in enumerate(countries):
    print(i, c)
  

### Advanced
#**********************
# Classes in Python
#**********************

#E14
# Declare 2 classes, and instantiate (declare one of each), and manipulate their variables
# Classes are the basis as a framework for object oriented programming

class someClass():
  def method1(self):
    print ("someClass method1")
    
  def method2(self, someString):
    print ("someClass method2: " + someString)
  
      
def someFunction():
  c = someClass()
  c.method1()
  c.method2("Foobar")
  
if __name__ == "__main__":
  someFunction()

File: module 2/part c/Answers-BeginnerToModerate.R
#**********************
# Variables in R
#**********************

#E1
# Declare a variable (i.e. something that stores a value) and initialize it (i.e. set something to that variable)
x <- 1:10  # Put the numbers 1-10 in the variable x

#print the variables
x

#E2
# Change the value of that variable (i.e. change what was initialized before), go ahead and try 
# to reset/declare, that variable a few times, check the variable type, (use type function, i.e. is 
# it int, string, etc.)
# Note: Run each variable declaration separately
x <- "foobar"
x <- "foo"
x <- 111
y <- c(5, 7, -1, 6, 7)

#(remember run each variable separately to see output)
x
y

# List/print objects
ls()  

#E3
# Global vs. local variables in functions
f <- function(a) {
   a=1
}

f(a)

# you should get Error: object 'a' not found is not defined' (variable ONLY exists in the function)
a

# This is the variable we declared earlier (it's GLOBAL)
x

#E4
#Now delete all the variables we made
rm(x)

#**********************
# Conditions in R
#**********************

#E5
# Write a switch statement in R
compare <- function(switchTo) {

switch (switchTo,
        "PicMe" = "blah",
        "NoPicMe" = "foobar",
        "OkFine" = "foo",
)
}

# Write an if/else statement in R
compare <- function() {
a = 10
b = 100 

  if(a < b){
    result= "a is less than b"
  }
  else if (a == b){
    result= "a is same as b"
  }
  else{
    result= "a is greater than b"
  }
  print(result)
}

compare()

#**********************
# Functions in R
#**********************

#E6
# Define a function (i.e. takes inputs, outputs, in this case a function with NO arguments,
# NO return values)
f <- function() {
}

f()

#E7
# Define a function (i.e. takes inputs, outputs, in this case a function with arguments)
# NO return values)
f <- function(a, b) {
   a+b
}

f(1,2)

#E8
# Define a function (i.e. takes inputs, outputs, in this case a function with arguments)
# return values)
f <- function(a, b, c) {
   result=a+b
   return(c(result,c))
}

f(1,2,3)

#E9
### Advanced
# Define a function (i.e. takes inputs, outputs, in this case a function with arguments)
# return values, has DEFAULT values and VARIABLE number of arguments)
f <- function(a, b = 1, c = 2, d = NULL, ...) {
   result=a+b

   return(c(result,c))
}

f(1,2,3,4, 2, 4, 2)

#**********************
# Loops in R
#**********************

#E10
# Create a while loop
f <- function() {
  
  # create a while loop
  i <- 1
  while (i < 4) {
   print(i)
   i = i+1
  }

  # create a for loop
  for (number in c(1,2,3,4)) { 
    print(paste("The number is", number)) 
  }
}

### Advanced
#**********************
# Classes in R
#**********************

#E11
# Declare 1 class, and instantiate (declare one of each), and manipulate their variables
# Classes are the basis within object oriented programming
# R has an S3, S4 class

#S3 class
someClass <- function(var=TRUE)
{

        me <- list(
                someMethod = var
       )

        class(me) <- append(class(me),"someClass")
        return(me)
}

someDefaultPerson <- someClass()

setSomeMethod <- function(elObjeto, newValue)
        {
                UseMethod("setSomeMethod",elObjeto)
        }

getSomeMethod <- function(elObjeto)
        {
                UseMethod("getSomeMethod",elObjeto)
        }

someDefaultPerson <- setSomeMethod(someDefaultPerson,FALSE)

#S4 class
someOtherClass <- setClass(
  "someOtherClass", 
      representation(someField = "character"), 
  
      prototype(someField = NA_character_)

  )

someOtherClass <- new("someOtherClass", someField = "foobar")

